File: src/utils/graphql.js
import { GRAPHQL_ENDPOINT } from '../config/constants.js';
import { json } from './response.js';

// Maximum concurrent GraphQL requests
const MAX_CONCURRENCY = 10;
let inflightCount = 0;
const inflightQueue = [];

// In-flight dedupe: Map<key, Promise>
const dedupeCache = new Map();
// Short-term response cache: Map<key, { ts, data }>
const shortTermCache = new Map();
const CACHE_TTL = 2_000;            // 2 seconds
const REQUEST_TIMEOUT_MS = 5_000;   // 5 seconds

/**
 * Race a promise against a timeout.
 */
async function withTimeout(promise, ms) {
  let timer;
  const timeout = new Promise((_, rej) =>
    timer = setTimeout(() => rej(new Error('Request timed out')), ms)
  );
  try {
    return await Promise.race([promise, timeout]);
  } finally {
    clearTimeout(timer);
  }
}

/** Wait until inflightCount < MAX_CONCURRENCY */
function waitForSlot() {
  if (inflightCount < MAX_CONCURRENCY) return Promise.resolve();
  return new Promise(resolve => inflightQueue.push(resolve));
}

/** Free up one slot and wake the next waiter */
function releaseSlot() {
  inflightCount--;
  if (inflightQueue.length) {
    const next = inflightQueue.shift();
    inflightCount++;
    next();
  }
}

/** Build a stable key from query+vars for dedupe/cache */
function canonicalKey(query, variables) {
  return JSON.stringify({ query, variables });
}

async function executeWithRetry(fn, retries = 1) {
  let lastError;
  for (let i = 0; i <= retries; i++) {
    try {
      return await fn();
    } catch (err) {
      lastError = err;
      console.warn(`GraphQL attempt ${i + 1} failed:`, err);
      // Optionally: only retry on specific errors
      if (err.status && err.status < 500) break; // don't retry 4xx
    }
  }
  throw lastError;
}

/**
 * Execute a GraphQL query with:
 * - short-term in-memory caching (2s TTL)
 * - in-flight deduplication
 * - max concurrency throttling
 * - global timeout (fetch+parse) of 5s
 */
export async function executeGraphQLQuery(
  query,
  variables = {},
  errorMessage = "GraphQL query failed"
) {
  const key = canonicalKey(query, variables);
  const now = Date.now();

  // 0) Short-term cache hit?
  const cached = shortTermCache.get(key);
  if (cached && (now - cached.ts) < CACHE_TTL) {
    return cached.data;
  }

  // 1) Deduplicate identical in-flight requests
  if (dedupeCache.has(key)) {
    return dedupeCache.get(key);
  }

  // 2) Throttle concurrency
  await waitForSlot();
  inflightCount++;

  // 3) Build the raw work (fetch + JSON parse)
  const rawPromise = (async () => {
    const response = await fetch(GRAPHQL_ENDPOINT, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ query, variables }),
    });

    if (!response.ok) {
      const text = await response.text();
      throw json(
        { error: errorMessage, status: response.status, details: text },
        { status: 502 }
      );
    }

    const data = await response.json();
    // cache success
    shortTermCache.set(key, { ts: Date.now(), data });
    return data;
  })();

  // 4) Enforce global timeout across fetch+parse
  const timedPromise = withTimeout(
    executeWithRetry(rawPromise, 1), // retry once
    REQUEST_TIMEOUT_MS
  );

  // 5) On completion or error, release slot & clear dedupe
  const finalPromise = timedPromise
    .catch(err => {
      if (err.message === 'Request timed out') {
        throw json(
          { error: `${errorMessage}: timed out after ${REQUEST_TIMEOUT_MS}ms` },
          { status: 504 }
        );
      }
      throw err;
    })
    .finally(() => {
      releaseSlot();
      dedupeCache.delete(key);
    });

  // 6) Remember in-flight
  dedupeCache.set(key, finalPromise);
  return finalPromise;
}

File: app/utils/graphql.py
"""
GraphQL utility for executing queries with caching, deduplication, and throttling
"""
import json
import time
import asyncio
from typing import Any, Dict, Optional
import httpx
from fastapi import HTTPException

from app.config.constants import GRAPHQL_ENDPOINT
from app.utils.response import json_response

# Maximum concurrent GraphQL requests
MAX_CONCURRENCY = 10
inflight_count = 0
inflight_queue = []

# In-flight dedupe: key -> Future
dedupe_cache = {}
# Short-term response cache: key -> {ts, data}
short_term_cache = {}
CACHE_TTL = 2_000  # 2 seconds
REQUEST_TIMEOUT_MS = 5_000  # 5 seconds

# Semaphore for controlling concurrency
semaphore = asyncio.Semaphore(MAX_CONCURRENCY)


def canonical_key(query: str, variables: Dict[str, Any]) -> str:
    """Build a stable key from query+vars for dedupe/cache"""
    return json.dumps({"query": query, "variables": variables}, sort_keys=True)


async def execute_with_retry(fn, retries=1, initial_delay=100):
    """Execute a function with retries and exponential backoff"""
    last_error = None
    delay = initial_delay

    for i in range(retries + 1):
        try:
            return await fn()
        except Exception as err:
            last_error = err
            print(f"GraphQL attempt {i + 1} failed: {err}")

            # Don't retry on 4xx errors
            if hasattr(err, 'status_code') and err.status_code < 500:
                break

            # Wait before retrying
            if i < retries:
                await asyncio.sleep(delay / 1000)  # Convert to seconds
            delay *= 2  # exponential backoff

    raise last_error


async def execute_graphql_query(
    query: str,
    variables: Dict[str, Any] = None,
    error_message: str = "GraphQL query failed"
) -> Dict[str, Any]:
    """
    Execute a GraphQL query with:
    - short-term in-memory caching (2s TTL)
    - in-flight deduplication
    - max concurrency throttling
    - global timeout (fetch+parse) of 5s
    """
    if variables is None:
        variables = {}
        
    key = canonical_key(query, variables)
    now = time.time() * 1000  # Convert to ms

    # 0) Short-term cache hit?
    if key in short_term_cache:
        cached = short_term_cache[key]
        if (now - cached["ts"]) < CACHE_TTL:
            return cached["data"]

    # 1) Deduplicate identical in-flight requests
    if key in dedupe_cache:
        return await dedupe_cache[key]

    # 2) Throttle concurrency using semaphore
    async def raw_work():
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(
                    GRAPHQL_ENDPOINT,
                    json={"query": query, "variables": variables},
                    headers={"Content-Type": "application/json"},
                    timeout=REQUEST_TIMEOUT_MS / 1000  # Convert to seconds
                )
                
                if not response.is_success:
                    text = response.text
                    raise HTTPException(
                        status_code=502,
                        detail={
                            "error": error_message,
                            "status": response.status_code,
                            "details": text
                        }
                    )

                data = response.json()

                if "errors" in data:
                    raise HTTPException(
                        status_code=502,
                        detail={
                            "error": error_message,
                            "status": 502,
                            "details": data["errors"]
                        }
                    )

                short_term_cache[key] = {"ts": time.time() * 1000, "data": data}
                return data
            except httpx.TimeoutException:
                raise HTTPException(
                    status_code=504,
                    detail={
                        "error": f"{error_message}: timed out after {REQUEST_TIMEOUT_MS}ms"
                    }
                )

    # 3) Execute with retry and timeout
    async with semaphore:
        try:
            # Create a task for the work
            task = asyncio.create_task(execute_with_retry(raw_work, 1))
            dedupe_cache[key] = task
            
            # Wait for the task to complete
            result = await task
            return result
        finally:
            # Clean up
            if key in dedupe_cache:
                del dedupe_cache[key]


async def sweep_graphql_cache(interval=30):
    while True:
        now = time.time() * 1000
        for k in [k for k,v in short_term_cache.items() if now - v["ts"] >= CACHE_TTL]:
            short_term_cache.pop(k, None)
        await asyncio.sleep(interval)